scanpipe/pipes/codebase.py

```diff
@@ -0,0 +1,83 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# http://nexb.com and https://github.com/nexB/scancode.io
+# The ScanCode.io software is licensed under the Apache License version 2.0.
+# Data generated with ScanCode.io is provided as-is without warranties.
+# ScanCode is a trademark of nexB Inc.
+#
+# You may not use this software except in compliance with the License.
+# You may obtain a copy of the License at: http://apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software distributed
+# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+#
+# Data Generated with ScanCode.io is provided on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, either express or implied. No content created from
+# ScanCode.io should be considered or used as legal advice. Consult an Attorney
+# for any legal advice.
+#
+# ScanCode.io is a free software code scanning tool from nexB Inc. and others.
+# Visit https://github.com/nexB/scancode.io for support and download.
+
+from django.core.exceptions import ObjectDoesNotExist
+
+from scanpipe.models import Project
+
+
+def sort_by_lower_name(resource):
+    return resource["name"].lower()
+
+
+def get_tree(resource, fields, codebase=None):
+    """
+    Return a tree as a dict structure starting from the provided `resource`.
+
+    The following classes are supported for the input `resource` object:
+     - scanpipe.models.CodebaseResource
+     - commoncode.resource.Resource
```

Comment:
That's great that you can support `commoncode.resource.Resource`. Yet I wonder if this may also be needed the other way too? e.g. feed a scancode plugin `process_codebase()` method a codebase object subclass or simialr that's behaving the exact same way as `commoncode.resource.Codebase` and that would yield objects that are behaving exactly like a `commoncode.resource.Resource` each being actually wrapping CodebaseResource instances.

An example of the `process_codebase()` is there https://github.com/nexB/scancode-toolkit/blob/develop/src/scancode/plugin_mark_source.py#L63

The benefit would be that scancode plugins could then be used as-is and fed a codebase that they process alright (and that could be filtered as some query set). This includes all the output plugins for instance.

Reply:
The `walk` is available on the ProjectCodebase for that purpose.
Also, on the CodebaseResource side, `is_file`, `is_dir`, `children`, `descendants` are implemented.
We can add more in the future as needed.